pkg/sql/plpgsql/parser/lexer.go

```diff
@@ -0,0 +1,482 @@
+	parenLevel := 0
+	for l.lastPos < len(l.tokens) {
+		tok := l.Peek()
+		if int(tok.id) == terminator1 && parenLevel == 0 {
+			terminatorMet = terminator1
+			break
+		} else if int(tok.id) == terminator2 && parenLevel == 0 {
+			terminatorMet = terminator2
+			break
+		} else if int(tok.id) == terminator3 && parenLevel == 0 {
+			terminatorMet = terminator3
+			break
+		} else if tok.id == '(' || tok.id == '[' {
+			parenLevel++
+
+		} else if tok.id == ')' || tok.id == ']' {
+			parenLevel--
+			if parenLevel < 0 {
+				panic(errors.AssertionFailedf("wrongly nested parentheses"))
+			}
+		}
+		exprTokenStrs = append(exprTokenStrs, tok.Str())
+		l.lastPos++
+	}
+	if parenLevel != 0 {
+		panic(errors.AssertionFailedf("parentheses is badly nested"))
+	}
+	if len(exprTokenStrs) == 0 {
+		//TODO(jane): show the terminator in the panic message.
+		l.setErr(errors.New("there should be at least one token for sql expression"))
+	}
+
+	return strings.Join(exprTokenStrs, " "), terminatorMet
+}
+
+func (l *lexer) ProcessQueryForCursorWithoutExplicitExpr(
+	openStmt *plpgsqltree.PLpgSQLStmtOpen,
+) {
+	l.lastPos++
+	if int(l.Peek().id) == EXECUTE {
+		dynamicQuery, endToken := l.ReadSqlExpressionStr2(USING, ';')
+		openStmt.DynamicQuery = dynamicQuery
+		if endToken == USING {
+			var expr string
+			for {
+				expr, endToken = l.ReadSqlExpressionStr2(',', ';')
+				openStmt.Params = append(openStmt.Params, expr)
+				if endToken != ',' {
+					break
+				}
+			}
+		}
+	} else {
+		openStmt.Query = l.ReadSqlExpressionStr(';')
+	}
+}
+
+// Peek peeks
+func (l *lexer) Peek() plpgsqlSymType {
+	if l.lastPos+1 < len(l.tokens) {
+		return l.tokens[l.lastPos+1]
+	}
+	return plpgsqlSymType{}
+}
+
+// PushBack move the lastP
+func (l *lexer) PushBack(n int) {
+	if l.lastPos-n >= 0 {
+		l.lastPos -= n
+	}
+}
+
+func (l *lexer) lastToken() plpgsqlSymType {
+	if l.lastPos < 0 {
+		return plpgsqlSymType{}
+	}
+
+	if l.lastPos >= len(l.tokens) {
+		return plpgsqlSymType{
+			id:  0,
+			pos: int32(len(l.in)),
+			str: "EOF",
+		}
+	}
+	return l.tokens[l.lastPos]
+}
+
+// SetStmt is called from the parser when the statement is constructed.
+func (l *lexer) SetStmt(stmt plpgsqltree.PLpgSQLStatement) {
+	l.stmt = stmt.(*plpgsqltree.PLpgSQLStmtBlock)
+}
+
+// PurposelyUnimplemented wraps Error, setting lastUnimplementedError.
+func (l *lexer) PurposelyUnimplemented(feature string, reason string) {
+	// We purposely do not use unimp here, as it appends hints to suggest that
+	// the error may be actively tracked as a bug.
+	l.lastError = errors.WithHint(
+		errors.WithTelemetry(
+			pgerror.Newf(pgcode.Syntax, "unimplemented: this syntax"),
+			fmt.Sprintf("sql.purposely_unimplemented.%s", feature),
+		),
+		reason,
+	)
+	l.populateErrorDetails()
+	l.lastError = &tree.UnsupportedError{
+		Err:         l.lastError,
+		FeatureName: feature,
+	}
+}
+
+// UnimplementedWithIssue wraps Error, setting lastUnimplementedError.
+func (l *lexer) UnimplementedWithIssue(issue int) {
```

Comment:
similar, looks like this is not used. We could always add it later, and also this looks identical to the ones in sql's lexer.

Reply:
Same comment for the other unused methods.